### DISCUSSION TITLE
Does Poweradmin allow for API settings?

### DISCUSSION THREAD

**Discussion by Jouwhost** [poweradmin/poweradmin#441] [Q&A]
We want renew our PowerDNS cluster and are wondering if there are any PowerDNS API settings build into Poweradmin.
We currently do not know how to create en API key in PowerDNS at all tbh. But also cannot find any screenshot or documentation if this is build into Poweradmin.
Would love to know if this is possible to do in Poweradmin and how to set this if Poweradmin does not have this.

**Comment by edmondas** [comment]
@Jouwhost Currently, Poweradmin does not support PowerDNS API keys. It works directly with SQL databases.

**Reply by hvisage** [reply]
A question did came up today: API access to PowerAdmin? 
(Use case: Kubernetes auto name insert/updates)

**Reply by edmondas** [reply]
@hvisage In the next release, there is a possibility that API will be the target, but it may require multiple releases to achieve a stable, well-designed interface. To update records now, you may be able to use a dynamic DNS update script: https://github.com/poweradmin/poweradmin/wiki/Dynamic-DNS

**Comment by edmondas** [comment]
Do you have any examples of typical use cases that you need?

For example PowerDNS-Admin API consists of two parts:
- The /powerdnsadmin endpoints manages PDA content (accounts, users, apikeys) and also allow domain creation/deletion
- The /server endpoints are proxying queries to the backend PowerDNS instance's API. PDA acts as a proxy managing several API Keys and permissions to the PowerDNS content.

https://github.com/PowerDNS-Admin/PowerDNS-Admin/blob/master/docs/API.md

Are you looking for something similar?

**Comment by RSpliet** [comment]
For me, the use-case would be LetsEncrypt auto-renewal for wildcard certs with my domain/DNS provider. Auto-renewal of wildcard certs can't be done with a HTTP-based authentication, Instead the renewal process requires the creation of a few TXT records, which can be automated through rfc2136. In order to have letsencrypt do that, I'd have to generate an API key with the same permissions as my user account. Being able to generate an API key from PowerAdmin myself would mean less work for my hosting provider.

Edit: Reading the original discussion point, it's slightly unclear to me on whether the focus of this discussion thread should be to allow _managing_ API keys or rather on PowerAdmin _using_ the PowerDNS APIs to do what it already does. In my case, I'm interested in the former, not the latter.

**Reply by edmondas** [reply]
I apologize for the delayed response - I've been focused on finalizing the current release, which will be published in the coming weeks. For your requested feature, I'm planning to create an initial API version in the upcoming 4.0.0 release that will expose a minimal set of endpoints. Since I'm not familiar with this standard, I'll need some time to understand how to implement it correctly. I would appreciate your assistance in testing the solution once it's developed.

**Reply by hvisage** [reply]
> the use-case would be LetsEncrypt auto-renewal for wildcard certs with my domain/DNS provider.

Do have a look at [acmedns](https://github.com/joohoi/acme-dns) as you can setup a CNAME and then use the client https://github.com/acme-dns/acme-dns-client to do the actual authentication

**Reply by edmondas** [reply]
Maybe it's time to gradually move to a Go-based solution in general? 😄 The API would be implemented in that language and run as a separate instance, which could be mapped to the /api URL through HTTP (nginx) or work in standalone mode. We could then incrementally migrate the old PHP code and change the license to be more permissive (like MIT).

**Reply by RSpliet** [reply]
@edmondas Appreciate the timeline you've provided. Sadly, I wouldn't be the right person to test this though. I don't have my own poweradmin deployment, rather I'm a customer with a registar+DNS provider (Exsilia) who allow me to manage my records using PowerAdmin. As such I don't have the set-up for testing unless/until they roll out the update.

**Comment by edmondas** [comment]
Created an issue to track implementation progress and document requirements (#627).

**Comment by edmondas** [comment]
I've added initial support for API and keys management. Core functionality is working, and I'm still finalizing endpoint exposure. This could be included in the 4.0.0 release.

![image](https://github.com/user-attachments/assets/6968b8e7-6821-41cc-8a50-a5dd7efb2ba1)

I'll complete the remaining endpoints before the release candidate, though I'm not sure if all use cases will be implemented in this iteration.

**Reply by platix** [reply]
great news.  Thank you very much.
It would be great if it could be limited per endpoint and per domain/registry to maintain the principle of least privilege.

**Comment by akops76** [comment]
I have setup my own powerdns/poweradmin instance and try to work with APIs. I use powerdns 5.0 and poweradmin 4.0.1 versions
First i enabled the powerdns api, and i have manage to add/delete records by using a custom made bash script. The problem here is security. The powerdns api key give full access to the user with the api key.
Then i  enabled the poweradmin api, and i also have manage to add/delete records by using another custom made bash script. Unfortunately same problem with security:
I have tried the following scenario:
a) Create a user with limited permissions to own and manage only one zone (ex test.com)
b) I temporary gave superuser permission the user. (only superusers have access to GUI API key menu).
c) Login with the user and create a new API KEY 
d) Remove superuser permissions from the user.
Then
a) Use the new API KEY to modify test.com zone. Modification was successful ( expected behavior).
b) Use the new API KEY to modify another zone of the server(eg test2.com).  Modification was also successful, which is wrong. Based on the user permissions his API KEY should not have the right to modify another zone on the server (except the test.com).

**Comment by edmondas** [comment]
@akops76 Thank you for taking the time to test the initial API version. I understand that access was limited to admins with full permissions, and you found a workaround to create an API key - though this key still retained full administrative power.

The current master branch now includes a new `api_manage_keys` permission that allows users to create API keys. However, granular control over what can be modified through these keys hasn't been implemented yet. The upcoming 4.1.0 release will include these security improvements.

Thank you for reporting this issue.

**Reply by edmondas** [reply]
Permission checks had been added in commit 7f01625. If any issues are found, please create new bug or feature reports.

**Comment by edmondas** [comment]
I'm closing this discussion. For more specific topics, please open new conversations or report issues. I've also created new tickets for implementing more granular API keys and audit logging - #795 and #796.
